client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedPusher.java

```diff
@@ -44,6 +44,73 @@
 
 public class SortBasedPusher extends MemoryConsumer {
 
+  class MemoryThresholdManager {
+
+    private long maxMemoryThresholdInBytes;
+    private long currentMemoryThresholdInBytes;
+
+    private double smallPushTolerateFactor;
+
+    MemoryThresholdManager(
+        int numPartitions,
+        long sendBufferSizeInBytes,
+        long currentMemoryThresholdInBytes,
+        double smallPushTolerateFactor) {
+      this.maxMemoryThresholdInBytes = numPartitions * sendBufferSizeInBytes;
+      this.currentMemoryThresholdInBytes = currentMemoryThresholdInBytes;
+      this.smallPushTolerateFactor = smallPushTolerateFactor;
+    }
+
+    private boolean shouldGrow() {
+      boolean enoughSpace = this.currentMemoryThresholdInBytes * 2 <= maxMemoryThresholdInBytes;
+      long shouldPushedBytes;
+      if (shouldPushedCount == 0) {
+        shouldPushedBytes = 0;
+      } else {
+        shouldPushedBytes =
+            (long) ((1 + smallPushTolerateFactor) * shouldPushedSizeInBytes / shouldPushedCount);
+      }
+      boolean tooManyPushed = pushedMemorySizeInBytes * 1.0 / pushedCount > shouldPushedBytes;
```

Comment:
`pushedMemorySizeInBytes * 1.0 / pushedCount` calculates average size of all pushes; `shouldPushedSizeInBytes / shouldPushedCount` calculates averages size of pushes where the size is close or equal to `pushBufferMaxSize`, so the later is almost aways bigger than the former, let alone multiply the later with (1 + `smallPushTolerateFactor`).

Maybe we can simplify the logic to check whether average size of all pushes is smaller than `pushBufferMaxSize / (1 + factor)` and grow if yes, and we don't need `shouldPushedSizeInBytes/shouldPushedCount`.

Reply:
oops, it is a typo, it should be (pushedMemorySizeInBytes * 1.0 / pushedCount * (1 + factor) < shouldPushedBytes

but I think you are right, we can just use pushBufferMaxSize

Reply:
I updated the code, however, I think we still need to keep the tracking of shouldPushedSizeInBytes/shouldPushedCount, otherwise it will grow threshold unnecessarily (you can also check my unit test )

e.g. if I have 32K buffer  and 256K data from the same partition, if we compare pushedBytes/pushedCounts with only buffer size (32K), we will still grow memory threshold since we would never be able to send exactly 32K for every push (because those extra bytes leading to that , at some point, we will have offset + record size > 32K), 

However,  the pushes in this case are expected instead of led by too many small partitions (i.e. the case we want to avoid), even we use hashpartitioner, we still need to do such a push...we **SHOULD NOT** grow threshold in this case since it won't help for anything

Reply:
I agree that for almost all cases the actual data pushed is smaller than `pushBufferMaxSize`, and that's why I suggested to compare with `pushBufferMaxSize/(1 + factor)`.